src/pymor/reductors/neural_network.py

```diff
@@ -240,6 +241,149 @@ def reconstruct(self, u):
+            The validation loss to reach during training. If `None`, the neural
+            network with the smallest validation loss is returned.
+        """
+
+        def __init__(self, fom, training_set, validation_set=None, validation_ratio=0.1,
+                     validation_loss=None):
+            assert 0 < validation_ratio < 1 or validation_set
+            self.__auto_init(locals())
+
+        def reduce(self, hidden_layers='[30, 30]', activation_function=torch.tanh,
+                   optimizer=optim.LBFGS, epochs=1000, batch_size=20, learning_rate=1.,
+                   restarts=10, seed=0):
+            """Reduce by training artificial neural networks.
+
+            Parameters
+            ----------
+            hidden_layers
+                Number of neurons in the hidden layers. Can either be fixed or
+                a Python expression string depending on the output size `O`
+                and the total dimension of the |Parameters| `P`.
+            activation_function
+                Activation function to use between the hidden layers.
+            optimizer
+                Algorithm to use as optimizer during training.
+            epochs
+                Maximum number of epochs for training.
+            batch_size
+                Batch size to use if optimizer allows mini-batching.
+            learning_rate
+                Step size to use in each optimization step.
+            restarts
+                Number of restarts of the training algorithm. Since the training
+                results highly depend on the initial starting point, i.e. the
+                initial weights and biases, it is advisable to train multiple
+                neural networks by starting with different initial values and
+                choose that one performing best on the validation set.
+            seed
+                Seed to use for various functions in PyTorch. Using a fixed seed,
+                it is possible to reproduce former results.
+
+            Returns
+            -------
+            rom
+                Reduced-order |NeuralNetworkOutputModel|.
+            """
+            assert restarts > 0
+            assert epochs > 0
+            assert batch_size > 0
+            assert learning_rate > 0.
+
+            # set a seed for the PyTorch initialization of weights and biases
+            # and further PyTorch methods
+            torch.manual_seed(seed)
+
+            # determine the numbers of neurons in the hidden layers
+            if isinstance(hidden_layers, str):
+                hidden_layers = eval(hidden_layers, {'O': self.fom.dim_output, 'P': self.fom.parameters.dim})
+            # input and output size of the neural network are prescribed by the
+            # dimension of the parameter space and the reduced basis size
+            assert isinstance(hidden_layers, list)
+            layer_sizes = self._compute_layer_sizes(hidden_layers)
+
+            # compute validation data
```

Comment:
There seems to be a lot of code duplication with the older reductor. Maybe it is possible to refactor that?

Reply:
I refactored the reductors [here](https://github.com/HenKlei/pymor/blob/refactor-ANN-reductors/src/pymor/reductors/neural_network.py), but I am actually not very happy with the result. Could you take a look @sdrave, please? I am open for any suggestion on how to improve this!

Reply:
I hope that I did not miss anything from our discussion today. It would be great if you could take a look at 72363c9, @sdrave. Tell me if there is something that should be changed.